### ISSUE TITLE
Components - default settings

### ISSUE THREAD

**Issue by m0jimo** [quasarframework/quasar#7379]
**Is your feature request related to a problem? Please describe.**
Most of my apps made by Quasar are targeted to touch devices. 
If you use `q-tooltip` it has a default tooltip position below the object. It is problematic on touch devices as it appears under your finger and tooltip is not visible.
 
You can easily define new position `<q-tooltip anchor="top middle" self="center middle">Mark as favorite</q-tooltip>` but you have to do it for each component.

**Describe the solution you'd like**
Would be nice to configure a default settings for some components not only `q-tooltip`. Other components for default settings could be `q-btn` or `q-input`.

I think it could work in the same way as it is now for e.g. `q-notify` `setDefaults` [Settings defaults](https://quasar.dev/quasar-plugins/notify#Setting-defaults).

**Describe alternatives you've considered**
I can always wrap Quasar component into my own and pass default properties.

**Comment by Evertvdw** [comment]
I would like to add my +1 to this, I am currently in the struggle to do this in a nice way but I can't find one. For instance the QTable component, I can create a wrapper that does basic styling of the component but I cannot use the slots of the QTable within the tags of the wrapper.  It would be nice if the defaults for all props of all components were defined somewhere in a quasar object which we can modify in a boot file or something like that.

**Comment by Demivan** [comment]
This is what we use in our project with v1 Quasar, but it is really hacky. Proper solution would be very nice to have.
```js
function setDefault (component, prop, type, value) {
    component.$options.props[prop] = { type: type, default: value }
  
    if (component.$options.propsData[prop] == null) {
        Vue.config.silent = true
        component[prop] = value
        Vue.config.silent = false
     }
}

const processed = {}

Vue.mixin({
    created () {
        // Only quasar components
        if (this.$options.name == null || !this.$options.name.startsWith('Q')) {
           return
        }

        // Do not initialize twice
        if (processed[this.$options.name]) {
            return
        }
  
        if (this.$options.name === 'QSelect') {
            setDefault(this, 'standout', [Boolean, String], 'bg-primary')
            setDefault(this, 'bgColor', [Boolean, String], 'primary')
        }

       ...
       
       processed[this.$options.name] = true
    })
}
```

**Comment by Demivan** [comment]
With Quasar 2 and Vue 3 there is no way to even hack it.
Are there any plans to support this?

**Comment by mJeromeDiaz** [comment]
You can just create a component with your tooltip configuration and call only that component.
I use this for dialog, frenchDate, circularProgress, and tooltip...

**Comment by Demivan** [comment]
Yeah, I solved it by creating wrapper components and changing Quasar autoimport code to import my components instead.

Still think there should be an easier way to configure Quasar defaults.

**Comment by Demivan** [comment]
Yeah, I solved it by creating wrapper components and changing Quasar autoimport code to import my components instead.

Still think there should be an easier way to configure Quasar defaults.

**Comment by zunnzunn** [comment]
Here's a good hack/workaround for Quasar2/Vue3 : create a [boot file](https://quasar.dev/quasar-cli-vite/boot-files#introduction) where you can set your own default props for specific Quasar components. For example, if you would like all of your `QInput` components to be `outlined` and `dense` per default:
```typescript
import { QInput } from "quasar"
import { boot } from "quasar/wrappers"

export default boot(({}) => {
     QInput.props.outlined = { type: Boolean, default: true }
     QInput.props.dense = { type: Boolean, default: true }
})
```

**Comment by pdanpdan** [comment]
I would go with something like this (to prevent having to fill everything in the prop):

```javascript
import { QInput } from 'quasar'
import { boot } from 'quasar/wrappers'

const defaultsQInput = {
  filled: true,
  dense: true,
  color: 'red',
  type: 'number',
  inputClass: 'text-right'
}

export default boot(() => {
  Object.keys(defaultsQInput).forEach(prop => {
    QInput.props[ prop ] = Array.isArray(QInput.props[ prop ]) === true || typeof QInput.props[ prop ] === 'function'
      ? { type: QInput.props[ prop ], default: defaultsQInput[ prop ] }
      : { ...QInput.props[ prop ], default: defaultsQInput[ prop ] }
  })
})
```

**Comment by yusufkandemir** [comment]
Here is a slightly different variant of @pdanpdan's approach, with some really basic TypeScript sprinkled on top:

```ts
import { QInput, QInputProps, QSelect, QSelectProps } from 'quasar';
import { boot } from 'quasar/wrappers';

const defaultsQInput: Partial<QInputProps> = {
  filled: true,
  dense: true,
  color: 'red',
  type: 'number',
  inputClass: 'text-right',
};

const defaultsQSelect: Partial<QSelectProps> = {
  filled: true,
  dense: true,
  inputClass: 'text-right',
};

export default boot(() => {
  Object.entries(defaultsQInput).forEach(([prop, value]) => {
    assignDefault(QInput.props, prop, value);
  });

  Object.entries(defaultsQSelect).forEach(([prop, value]) => {
    assignDefault(QSelect.props, prop, value);
  });
});

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function assignDefault(props: any, prop: string, value: any) {
  /* eslint-disable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access */
  props[prop] =
    Array.isArray(props[prop]) === true || typeof props[prop] === 'function'
      ? { type: props[prop], default: value }
      : { ...props[prop], default: value };
  /* eslint-enable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access */
}
```

**Comment by RPainter8West** [comment]
Thanks for the above @yusufkandemir and everyone here.  This is my version which is a bit more terse, but still has type checking on the props.

```typescript
import { QInput, QInputProps, QSelect, QSelectProps } from "quasar";
import { boot } from "quasar/wrappers";

export default boot(() => {
  SetComponentDefaults<QInputProps>(QInput, {
    outlined: true,
    dense: true,
    stackLabel: true,
  });
  SetComponentDefaults<QSelectProps>(QSelect, {
    outlined: true,
    dense: true,
    stackLabel: true,
  });
});

/**
 * Set some default properties on a component
 */
const SetComponentDefaults = <T>(component: any, defaults: Partial<T>): void => {
  /* eslint-disable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access */
  Object.keys(defaults).forEach((prop: string) => {
    component.props[prop] =
      Array.isArray(component.props[prop]) === true || typeof component.props[prop] === "function"
        ? { type: component.props[prop], default: (defaults as Record<string, any>)[prop] }
        : { ...component.props[prop], default: (defaults as Record<string, any>)[prop] };
  });
  /* eslint-enable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access */
};
```

**Comment by cluidold** [comment]
Hi!

This is a really nice way to set default via boot file. But is there any solution for setting defaults at runtime (e.g. via a settings page and some switches)?

**Comment by ViToni** [comment]
This approach works nice and helps reduce clutter and makes settings more consistent all over. However I'm having issues with properties which haven themselves properties:

```js
import { QTable } from "quasar";
import { boot } from "quasar/wrappers";

const defaultsQTable = {
    pagination: {
        rowsPerPage: 50
    },
    "rows-per-page-options": [20, 50, 100, 200, 500, 0]
};

function mapDefaultsToObject(defaultProps, objectType) {
    Object.keys(defaultProps).forEach(prop => {
        objectType.props[prop] =
        Array.isArray(objectType.props[prop]) === true || typeof objectType.props[prop] === "function"
            ? { type: objectType.props[prop], default: defaultProps[prop] }
            : { ...objectType.props[prop], default: defaultProps[prop] };
    });
}

export default boot(() => {
    mapDefaultsToObject(defaultsQTable, QTable);
});
```

While `rows-per-page-options` gets set, `pagination` doesn't seem to have been set as default.

**Comment by pdanpdan** [comment]
`pagination` is object, so the default should be a function returning the default

```js
function mapDefaultsToObject(defaultProps, objectType) {
  Object.keys(defaultProps).forEach(prop => {
    const default = defaultProps[prop] === Object(defaultProps[prop])
      ? () => defaultProps[prop]
      : defaultProps[prop]
    objectType.props[prop] = Array.isArray(objectType.props[prop]) === true || typeof objectType.props[prop] === 'function'
      ? { type: objectType.props[prop], default }
      : { ...objectType.props[prop], default }
  })
}
```

**Comment by ViToni** [comment]
@pdanpdan Mulţumesc frumos pentru răspunsul repede.
Unfortunately it doesn't seem to work for `pagination.rowsPerPage`. As I'm pretty fresh in regards of the JavaScript landscape I'm wondering if it might be related due to this specific property being calculated rather than just instantiated.
Or so it seems to me from: https://github.com/quasarframework/quasar/blob/dev/ui/src/components/table/table-pagination.js#L35-L44

**Comment by prashantnirgun** [comment]
> Thanks for the above @yusufkandemir and everyone here. This is my version which is a bit more terse, but still has type checking on the props.
> 
> ```ts
> import { QInput, QInputProps, QSelect, QSelectProps } from "quasar";
> import { boot } from "quasar/wrappers";
> 
> export default boot(() => {
>   SetComponentDefaults<QInputProps>(QInput, {
>     outlined: true,
>     dense: true,
>     stackLabel: true,
>   });
>   SetComponentDefaults<QSelectProps>(QSelect, {
>     outlined: true,
>     dense: true,
>     stackLabel: true,
>   });
> });
> 
> /**
>  * Set some default properties on a component
>  */
> const SetComponentDefaults = <T>(component: any, defaults: Partial<T>): void => {
>   /* eslint-disable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access */
>   Object.keys(defaults).forEach((prop: string) => {
>     component.props[prop] =
>       Array.isArray(component.props[prop]) === true || typeof component.props[prop] === "function"
>         ? { type: component.props[prop], default: (defaults as Record<string, any>)[prop] }
>         : { ...component.props[prop], default: (defaults as Record<string, any>)[prop] };
>   });
>   /* eslint-enable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access */
> };
> ```

I am not using typescript I try to remove the type definition but did not succeed now its giving the error can any one fix it
```
[{
	"resource": "/home/prashant/www/apmc/apmc-quasar/src/boot/cd.ts",
	"owner": "eslint",
	"severity": 8,
	"message": "Parsing error: Unexpected token <",
	"source": "eslint",
	"startLineNumber": 20,
	"startColumn": 30,
	"endLineNumber": 20,
	"endColumn": 30
}] 
```

**Comment by ThanhNg** [comment]
@rstoenescu Why did this get closed as completed? So how can we achieve this, for vite plugin or umd?

**Comment by yusufkandemir** [comment]
@ThanhNg for vite-plugin, you can place the same code except the `export default boot(() => {` part in `main.js` or equivalent. For UMD, on top of that, you have to remove imports and reference the components as `Quasar.QInput`.

**Comment by Duckiduc** [comment]
> Thanks for the above @yusufkandemir and everyone here. This is my version which is a bit more terse, but still has type checking on the props.
> 
> ```ts
> import { QInput, QInputProps, QSelect, QSelectProps } from "quasar";
> import { boot } from "quasar/wrappers";
> 
> export default boot(() => {
>   SetComponentDefaults<QInputProps>(QInput, {
>     outlined: true,
>     dense: true,
>     stackLabel: true,
>   });
>   SetComponentDefaults<QSelectProps>(QSelect, {
>     outlined: true,
>     dense: true,
>     stackLabel: true,
>   });
> });
> 
> /**
>  * Set some default properties on a component
>  */
> const SetComponentDefaults = <T>(component: any, defaults: Partial<T>): void => {
>   /* eslint-disable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access */
>   Object.keys(defaults).forEach((prop: string) => {
>     component.props[prop] =
>       Array.isArray(component.props[prop]) === true || typeof component.props[prop] === "function"
>         ? { type: component.props[prop], default: (defaults as Record<string, any>)[prop] }
>         : { ...component.props[prop], default: (defaults as Record<string, any>)[prop] };
>   });
>   /* eslint-enable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access */
> };
> ```

For those using Typescript here is what I've done to prevent the use of the `any` type:
```ts
import { ComponentConstructor } from 'quasar';

const SetComponentDefaults = <T>(component: ComponentConstructor<T>, defaults: Partial<T>): void => {
  /* eslint-disable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access */
  (Object.keys(defaults) as (keyof typeof defaults)[]).forEach((prop: keyof typeof defaults) => {
    component.props[prop] =
      Array.isArray(component.props[prop]) === true || typeof component.props[prop] === 'function'
        ? {
            type: component.props[prop],
            default: defaults[prop],
          }
        : {
            ...component.props[prop],
            default: defaults[prop],
          };
  });
  /* eslint-enable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access */
};
```

**Comment by ivanjaros** [comment]
Setting default props is working with this. But how about attributes, like class? For example, I want to have white tooltips with dark text instead of dark background and white text. There does not seem to be a way to achieve this. And with tooltip, there is no prop to set, it has to be on the element itself which is just no go due to sheer amount of them in the application.
